pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -369,7 +369,9 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
 	// Examine each replica with respect to its ordinal
 	for i := range replicas {
 		// delete and recreate failed pods
-		if isFailed(replicas[i]) {
+		// We can count pending pods as failed pods for the purposes
+		// of a stateful set
+		if isFailed(replicas[i]) || isStatefullyStuck(set, replicas[i], currentRevision, updateRevision) {
```

Comment:
I reproduced this problem in another PR https://github.com/kubernetes/kubernetes/pull/78182 and found out this line of code causes the stuck problem. But update the condition of stateful set pod deletion should be very careful.